rust/xaynet-desktop/src/participant.rs

```diff
@@ -0,0 +1,192 @@
+    }
+}
+
+impl Notify for Notifier {
+    fn new_round(&mut self) {
+        self.notify(Event::NewRound)
+    }
+    fn sum(&mut self) {
+        self.notify(Event::Sum)
+    }
+    fn update(&mut self) {
+        self.notify(Event::Update)
+    }
+    fn load_model(&mut self) {
+        self.notify(Event::LoadModel)
+    }
+    fn idle(&mut self) {
+        self.notify(Event::Idle)
+    }
+}
+
+/// A store shared between by the participant and its internal state machine. When the
+/// state machine emits a [`Event::LoadModel`] event, the participant is expected to
+/// load its model into the store. See [`Participant::set_model()`].
+#[derive(Clone)]
+struct Store(Arc<Mutex<Option<Model>>>);
+
+impl Store {
+    /// Create a new model store.
+    fn new() -> Self {
+        Self(Arc::new(Mutex::new(None)))
+    }
+}
+
+#[async_trait]
+impl ModelStore for Store {
+    type Model = Model;
+    type Error = std::convert::Infallible;
+
+    async fn load_model(&mut self) -> Result<Option<Self::Model>, Self::Error> {
+        Ok(self.0.lock().unwrap().take())
+    }
+}
+
+pub struct Participant {
+    /// Receiver for the events emitted by the state machine
+    _events: Events,
+    /// Model store where the participant should load its model, when
+    /// `self.should_set_model` is `true`.
+    store: Store,
+    ///
+    thread: thread::JoinHandle<()>,
+
+    shutdown: tokio::sync::mpsc::Sender<()>,
+}
+
+#[derive(Error, Debug)]
+pub enum InitError {
+    #[error("failed to deserialize the participant state {:?}", _0)]
+    Deserialization(#[from] Box<bincode::ErrorKind>),
+    #[error("failed to initialize the participant runtime {:?}", _0)]
+    Runtime(std::io::Error),
+    #[error("failed to initialize HTTP client {:?}", _0)]
+    Client(#[from] ClientError),
+    #[error("invalid participant settings {:?}", _0)]
+    InvalidSettings(#[from] SettingsError),
+}
+
+impl Participant {
+    /// Create a new participant with the given settings
+    pub fn new() -> Result<Self, InitError> {
+        let env_filter = EnvFilter::try_from_env("XAYNET_CLIENT");
+        if env_filter.is_ok() {
+            let _fmt_subscriber = FmtSubscriber::builder()
+                .with_env_filter(env_filter.unwrap())
+                .with_ansi(true)
+                .init();
+        }
+
+        let mut settings = Settings::new();
+        settings.set_keys(xaynet_core::crypto::SigningKeyPair::generate());
+        settings.set_scalar(1.0);
+        settings.set_url(String::from("http://127.0.0.1:8081"));
+
+        let (url, pet_settings) = settings.try_into()?;
+        let client = new_client(url.as_str(), None, None)?;
+        let (events, notifier) = Events::new();
+        let store = Store::new();
+        let state_machine = StateMachine::new(pet_settings, client, store.clone(), notifier);
+        let (shutdown_tx, mut shutdown_rx) = tokio::sync::mpsc::channel::<()>(1);
+
+        let handle = thread::spawn(move || {
```

Comment:
I think it would be nice to do this in a separate `start` method: `new` instantiate the object, but should start a background task imo.